qiskit/providers/experiment/experiment_data.py

```diff
@@ -0,0 +1,674 @@
+from .analysis_result import AnalysisResultV1 as AnalysisResult
+from .json import NumpyEncoder, NumpyDecoder
+from .utils import save_data
+
+LOG = logging.getLogger(__name__)
+
+
+def auto_save(func: Callable):
+    """Decorate the input function."""
+    @wraps(func)
+    def _wrapped(self, *args, **kwargs):
+        return_val = func(self, *args, **kwargs)
+        if self.auto_save:
+            self.save()
+        return return_val
+    return _wrapped
+
+
+class ExperimentData:
+    """Base common type for all versioned ExperimentData classes.
+
+    Note this class should not be inherited from directly, it is intended
+    to be used for type checking. When implementing a provider you should use
+    the versioned abstract classes as the parent class and not this class
+    directly.
+    """
+    version = 0
+
+
+class ExperimentDataV1(ExperimentData):
+    """Class to handle experiment data.
+
+    This class serves as a data container for experiment data and any analysis
+    results or figures associated with the experiment. It also provides methods
+    used to interact with the database, such as saving the experiment data in
+    the database or retrieving its figured.
+    """
+
+    version = 1
+    _metadata_version = 1
+    _executor = futures.ThreadPoolExecutor()
+    """Threads used for asynchronous processing."""
+
+    _json_encoder = NumpyEncoder
+    _json_decoder = NumpyDecoder
+
+    def __init__(
+            self,
+            backend: Union[Backend, BaseBackend],
+            experiment_type: str,
+            experiment_id: Optional[str] = None,
+            tags: Optional[List[str]] = None,
+            job_ids: Optional[List[str]] = None,
+            share_level: Optional[str] = None,
+            metadata: Optional[Dict] = None,
+            figure_names: Optional[List[str]] = None,
+            notes: Optional[str] = None
+    ):
+        """Initializes the experiment data.
+
+        Args:
+            backend: Backend the experiment runs on. It can either be a
+                :class:`~qiskit.providers.Backend` instance or just backend name.
+            experiment_type: Experiment type.
+            experiment_id: Experiment ID. One will be generated if not supplied.
+            tags: Tags to be associated with the experiment.
+            job_ids: IDs of jobs submitted for the experiment.
+            share_level: Whether this experiment can be shared with others. This
+                is applicable only if the experiment service supports sharing. See
+                the specific service provider's documentation on valid values.
+            metadata: Additional experiment metadata.
+            figure_names: Name of figures associated with this experiment.
+            notes: Freeform notes about the experiment.
+
+        Raises:
+            ExperimentError: If an input argument is invalid.
+        """
+        metadata = metadata or {}
+        self._source = metadata.pop(
```

Comment:
Use `get` instead of `pop`; we need to be cautious with mutable objects, like dictionaries, when they are passed as function parameters, and refrain from modifying them.

Reply:
fix in 42833a9